fblock.go

```diff
@@ -0,0 +1,455 @@
+	fb.Header.PrevKeyMR = new(Bytes32)
+	i += copy(fb.Header.PrevKeyMR[:], data[i:i+len(fb.Header.PrevKeyMR)])
+	fb.Header.PrevLedgerKeyMR = new(Bytes32)
+	i += copy(fb.Header.PrevLedgerKeyMR[:], data[i:i+len(fb.Header.PrevLedgerKeyMR)])
+	fb.Header.ExchangeRate = binary.BigEndian.Uint64(data[i : i+8])
+	i += 8
+	fb.Header.Height = binary.BigEndian.Uint32(data[i : i+4])
+	i += 4
+
+	expansionsize, read := binary.Varint(data[i:])
+	fb.Header.ExpansionSize = expansionsize
+	i += read
+	fb.Header.ExpansionBytes = make([]byte, expansionsize)
+	copy(fb.Header.ExpansionBytes, data[i:i+int(expansionsize)])
+	i += int(expansionsize) // TODO: This should be safe, as the expansion size is never > max int
+
+	fb.Header.TransactionCount = binary.BigEndian.Uint32(data[i : i+4])
+	i += 4
+	fb.Header.BodySize = binary.BigEndian.Uint32(data[i : i+4])
+	i += 4
+
+	// TODO: Body
+	fb.Transactions = make([]*FactoidTransaction, fb.Header.TransactionCount)
+	period := 0
+	for c := range fb.Transactions {
+		// Before each fct tx, we need to see if there is a marker byte that indicates a minute marker
+		for data[i] == FBlockMinuteMarker {
+			if period > len(fb.endOfPeriod) {
+				return fmt.Errorf("too many minute markers")
+			}
+			fb.endOfPeriod[period] = c
+			period++ // The next period encountered will be the next minute
+			i += 1
+		}
+
+		fb.Transactions[c] = new(FactoidTransaction)
+		read, err := fb.Transactions[c].Decode(data[i:])
+		if err != nil {
+			return err
+		}
+		i += read
+	}
+
+	// If we have not hit the end of our periods, a single byte will remain
+	for period < len(fb.endOfPeriod) {
+		fb.endOfPeriod[period] = int(fb.Header.TransactionCount)
+		period++ // The next period encountered will be the next minute
+		i += 1
+	}
+
+	// Set out computed fields
+	keyMr, err := fb.ComputeKeyMR()
+	if err != nil {
+		return err
+	}
+	ledgerMr, err := fb.ComputeLedgerKeyMR()
+	if err != nil {
+		return err
+	}
+
+	fb.KeyMR = &keyMr
+	fb.LedgerKeyMR = &ledgerMr
+
+	return nil
+}
+
+func (fb *FBlock) MarshalBinary() ([]byte, error) {
+	// Header has variable size
+	header, err := fb.MarshalBinaryHeader() // This checks for populated
+	if err != nil {
+		return nil, err
+	}
+
+	data := make([]byte, len(header)+int(fb.Header.BodySize))
+
+	i := 0
+	i += copy(data[i:], header)
+
+	period := 0
+	for c, transaction := range fb.Transactions {
+		for period < len(fb.endOfPeriod) && // If minute marked remain to be written
+			fb.endOfPeriod[period] > 0 && // If the period markers are actually set (ignore otherwise)
+			c == fb.endOfPeriod[period] { // This TX is the market point
+			data[i] = FBlockMinuteMarker
+			i += 1
+			period++
+		}
+
+		tData, err := transaction.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		i += copy(data[i:], tData)
+	}
+
+	for period < len(fb.endOfPeriod) {
+		data[i] = FBlockMinuteMarker
+		i += 1
+		period++
+	}
+
+	return data, nil
+}
+
+func (fb *FBlock) MarshalBinaryHeader() ([]byte, error) {
```

Comment:
Because of my change to how and when I'm computing KeyMRs and ChainIDs, I no longer needed to have the MarshalBinaryHeader function to be separate from MarshalBinary. I'd like to see the same changes reflected here eventually. Please just move this into MarshalBinary. Please also then use a single allocation for data if the exact full size can be known.

Reply:
The exact size of the header is known, but not the body. I'd like to keep this separated as I use the MarshalBinaryHeader() in the KeyMr code, since the header is a branch and the body is the other branch. So this function is used outside MarshalBinary

Reply:
I see what you mean for the changes.
I can move this in, I just like the code separation.